### PR TITLE
[Incremental] Use source file fingerprints *without* top-level type bodies for incremental imports

### PR DESCRIPTION
**Pull request by davidungar**
Has been rebased atop main after https://github.com/apple/swift/pull/36168 merges.

Add `InterfaceHashExcludingTypeMembers` to `BasicSourceFileInfo` and use it for module fingerprints in order to make incremental imports work.

Also adds a test.

### CODE REVIEW

**Comment by davidungar** [comment]
@swift-ci please smoke test macos platform

File: lib/Serialization/ModuleFile.cpp [hunk 1]
```diff
@@ -992,19 +999,26 @@ void ModuleFile::collectBasicSourceFileInfo(
...
+      llvm::errs() << "Unconvertible fingerprint excluding type members'"
+                   << fpStrExcludingTypeMembers << "'\n";
+      abort();
+    }
+    callback(BasicSourceFileInfo(filePath,
+                                 fingerprintIncludingTypeMembers.getValue(),
+                                 fingerprintExcludingTypeMembers.getValue(),
+                                 llvm::sys::TimePoint<>(std::chrono::nanoseconds(timestamp)),
+                                 fileSize);
+             }
```
Comment (rintaro) [inline lib/Serialization/ModuleFile.cpp:1021]: nitpick:
```suggestion
  }
```

Comment (davidungar) [reply lib/Serialization/ModuleFile.cpp:1021]: Thank you! Will fix.

**Comment by davidungar** [comment]
@swift-ci please smoke test

**Comment by davidungar** [comment]
@swift-ci please smoke test and merge

**Pull request merged**